git-protocol/src/upload_pack.rs

```diff
@@ -0,0 +1,134 @@
+                Err("empty path")
+            } else {
+                Ok(path.to_owned())
+            }
+        })?;
+        let host = match parts.next() {
+            None | Some("") => None,
+            Some(host) => match host.strip_prefix("host=") {
+                None => return Err("invalid host"),
+                Some(host) => match host.split_once(':') {
+                    None => Some((host.to_owned(), None)),
+                    Some((host, port)) => {
+                        let port = port.parse::<u16>().or(Err("invalid port"))?;
+                        Some((host.to_owned(), Some(port)))
+                    },
+                },
+            },
+        };
+        let extra = parts
+            .skip_while(|part| part.is_empty())
+            .map(|part| match part.split_once('=') {
+                None => (part.to_owned(), None),
+                Some((k, v)) => (k.to_owned(), Some(v.to_owned())),
+            })
+            .collect();
+
+        Ok(Self { path, host, extra })
+    }
+}
+
+pub async fn upload_pack<R, W>(
+    git_dir: impl AsRef<Path>,
+    recv: R,
+    mut send: W,
+) -> io::Result<(Header, impl Future<Output = io::Result<ExitStatus>>)>
+where
+    R: AsyncRead + Unpin,
+    W: AsyncWrite + Unpin,
+{
+    let mut recv = git_packetline::StreamingPeekableIter::new(recv, &[]);
+    let header: Header = {
+        let pkt = recv
+            .read_line()
+            .await
+            .ok_or_else(|| invalid_data("missing header"))?
+            .map_err(invalid_data)?
+            .map_err(invalid_data)?;
+        match pkt {
+            PacketLine::Data(data) => std::str::from_utf8(data)
+                .map_err(invalid_data)?
+                .parse()
+                .map_err(invalid_data),
+            _ => Err(invalid_data("not a header packet")),
+        }
+    }?;
+    let namespace = header.path.clone();
+    let mut recv = recv.into_inner();
+
+    let fut = async move {
+        let mut child = Command::new("git")
+            .current_dir(git_dir)
+            .env_clear()
+            .envs(std::env::vars().filter(|(key, _)| key == "PATH" || key.starts_with("GIT_TRACE")))
+            .env("GIT_PROTOCOL", "version=2")
+            .args(&[
+                "-c",
+                "uploadpack.allowanysha1inwant=true",
+                "-c",
+                "uploadpack.allowrefinwant=true",
+                "-c",
+                "lsrefs.unborn=ignore",
+                &format!("--namespace={}", namespace),
+                "upload-pack",
+                "--strict",
+                "--timeout=5",
+                ".",
+            ])
+            .stdout(Stdio::piped())
+            .stdin(Stdio::piped())
+            .stderr(Stdio::inherit())
+            .kill_on_drop(true)
+            .reap_on_drop(true)
+            .spawn()?;
+
+        let mut stdin = child.stdin.take().unwrap();
+        let mut stdout = child.stdout.take().unwrap();
+
+        try_join!(
```

Comment:
If this would be done with [Stream::collect](https://docs.rs/futures-lite/1.12.0/futures_lite/stream/trait.StreamExt.html#method.try_collect) the `futures-util` dependency could be removed.

It _should_ be the same.

Reply:
I don’t think that’s the same: `try_join!` drives all futures in a loop, while `collect` only the head of the stream. Technically, we would even want `FuturesUnordered` or `select!`.

Reply:
I had found https://github.com/smol-rs/futures-lite/issues/6, but that seems to have fizzled. Will need to contemplate if I like `try_zip(a, try_zip(b, c))` (although we use something like this elsewhere).

Reply:
Additionally I think it's fair asking the value question, because even if `futures-util` would be removed here it's likely to be in a lot of other places within the project.
As a disclaimer, I might have a bias trying to replace `futures-util` with `futures-lite` to save 0.5s of compilation time, but it's clearly not worth it 😅.

Background: When working on an Intel Macbook Air I put a lot of effort into optimizing compile times of `prodash`, which now works without `futures-util`. That time seems to be stuck with me.